File: tests/rosetta/transpiler/Java/even-or-odd.java
public class Main {

    static java.math.BigInteger parseBigInt(String str) {
        int i = 0;
        boolean neg = false;
        if (_runeLen(str) > 0 && (_substr(str, 0, 1).equals("-"))) {
            neg = true;
            i = 1;
        }
        java.math.BigInteger n = java.math.BigInteger.valueOf(0);
        while (i < _runeLen(str)) {
            String ch = _substr(str, i, i + 1);
            int d = Integer.parseInt(ch);
            n = n.multiply((java.math.BigInteger.valueOf(10))).add((new java.math.BigInteger(String.valueOf(d))));
            i = i + 1;
        }
        if (neg) {
            n = (n).negate();
        }
        return n;
    }

    static String pad(int n, int width) {
        String s = (String)(_p(n));
        while (_runeLen(s) < width) {
            s = " " + s;
        }
        return s;
    }

    static void showInt(int n) {
        String line = "Testing integer " + String.valueOf(pad(n, 3)) + ":  ";
        if (Math.floorMod(n, 2) == 0) {
            line = line + "even ";
        } else {
            line = line + " odd ";
        }
        if (Math.floorMod(n, 2) == 0) {
            line = line + "even";
        } else {
            line = line + " odd";
        }
        System.out.println(line);
    }

    static void showBig(String s) {
        java.math.BigInteger b = parseBigInt(s);
        String line_1 = "Testing big integer " + (String)(_p(b)) + ":  ";
        if (b.remainder((java.math.BigInteger.valueOf(2))).compareTo(java.math.BigInteger.valueOf(0)) == 0) {
            line_1 = line_1 + "even";
        } else {
            line_1 = line_1 + "odd";
        }
        System.out.println(line_1);
    }

    static void main() {
        showInt(-2);
        showInt(-1);
        showInt(0);
        showInt(1);
        showInt(2);
        showBig("-222222222222222222222222222222222222");
        showBig("-1");
        showBig("0");
        showBig("1");
        showBig("222222222222222222222222222222222222");
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            main();
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static int _runeLen(String s) {
        return s.codePointCount(0, s.length());
    }

    static String _substr(String s, int i, int j) {
        int start = s.offsetByCodePoints(0, i);
        int end = s.offsetByCodePoints(0, j);
        return s.substring(start, end);
    }

    static String _p(Object v) {
        return v != null ? String.valueOf(v) : "<nil>";
    }
}

File: tests/rosetta/transpiler/FS/even-or-odd.fs
// Generated 2025-08-04 00:30 +0700

exception Return
let mutable __ret = ()

let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _substring (s:string) (start:int) (finish:int) =
    let len = String.length s
    let mutable st = if start < 0 then len + start else start
    let mutable en = if finish < 0 then len + finish else finish
    if st < 0 then st <- 0
    if st > len then st <- len
    if en > len then en <- len
    if st > en then st <- en
    s.Substring(st, en - st)

let rec parseBigInt (str: string) =
    let mutable __ret : bigint = Unchecked.defaultof<bigint>
    let mutable str = str
    try
        let mutable i: int = 0
        let mutable neg: bool = false
        if ((String.length (str)) > 0) && ((_substring str 0 1) = "-") then
            neg <- true
            i <- 1
        let mutable n: bigint = bigint 0
        while i < (String.length (str)) do
            let ch: string = _substring str i (i + 1)
            let d: int = int ch
            n <- (n * (bigint 10)) + (bigint d)
            i <- i + 1
        if neg then
            n <- -n
        __ret <- n
        raise Return
        __ret
    with
        | Return -> __ret
and pad (n: int) (width: int) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable n = n
    let mutable width = width
    try
        let mutable s: string = string (n)
        while (String.length (s)) < width do
            s <- " " + s
        __ret <- s
        raise Return
        __ret
    with
        | Return -> __ret
and showInt (n: int) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable n = n
    try
        let mutable line: string = ("Testing integer " + (pad (n) (3))) + ":  "
        if (((n % 2 + 2) % 2)) = 0 then
            line <- line + "even "
        else
            line <- line + " odd "
        if (((n % 2 + 2) % 2)) = 0 then
            line <- line + "even"
        else
            line <- line + " odd"
        printfn "%s" (line)
        __ret
    with
        | Return -> __ret
and showBig (s: string) =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    let mutable s = s
    try
        let b: bigint = parseBigInt (s)
        let mutable line: string = ("Testing big integer " + (string (b))) + ":  "
        if (((b % (bigint 2) + (bigint 2)) % (bigint 2))) = (bigint 0) then
            line <- line + "even"
        else
            line <- line + "odd"
        printfn "%s" (line)
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        showInt (-2)
        showInt (-1)
        showInt (0)
        showInt (1)
        showInt (2)
        showBig ("-222222222222222222222222222222222222")
        showBig ("-1")
        showBig ("0")
        showBig ("1")
        showBig ("222222222222222222222222222222222222")
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
